### PR TITLE
Add more flexible fork choice scoring

### LINKED ISSUE #782: LMD GHOST scoring is broken
### What is wrong?

See https://github.com/ethereum/trinity/pull/685/files/b4dbf6a58d444578ee1b882ca21256eaf1a5f6a4#r294035068 on issue #685.

The score is the sum of the attestation data and the block root data but this obscures the semantics of the fork choice scoring.

The attestation data yields a balance and only upon ties do we want to consider the block root data. The code in #685 just sums them which could result in mis-scoring a block where the balances are close and the block root puts the incorrect block as having a higher score.

### How can it be fixed

We need a more flexible score type for the fork choice rule (including a way to {de,}serialize from disk).

1. The `ForkChoiceScoringFn` type should return an opaque `Score` type (instead of just an `int`).
2. Different fork choice scorings can provide different implementations of `Score` as makes sense -- the main thing `Score` needs is to satisfy a comparator interface and a {de,}serializing ability for writing scores to the DB.

something like:

```python
class Score(abc.ABC):
    @abstractmethod
    def cmp(self, other: Score) -> int:
        """ Return [-1, 0, 1] if ``self`` is [less than, equal to, greater than] ``other``."""
        pass

    @abstractmethod
    def serialize(self) -> bytesOrSomething:
        pass

    @abstractmethod
    @classmethod
    def deserialize(cls, contents: bytesOrSomething) -> Score:
        pass
```

### PR DESCRIPTION
**Pull request by ralexstokes**
### What was wrong?

Fixes #782.

### How was it fixed?

Introduces an opaque "score" type so that different forks of the protocol can specify their own format for scoring a block tree when computing the fork choice.

#### Cute Animal Picture

put a cute animal picture link inside the parentheses

### CODE REVIEW

**Comment by ralexstokes** [comment]
CI failures seem unrelated to this PR

File: eth2/beacon/chains/base.py [hunk 1]
```diff
@@ -236,21 +237,13 @@ def from_genesis(
...
-        genesis_config: Eth2GenesisConfig,
-    ) -> "BaseBeaconChain":
-        """
-        Initialize the ``BeaconChain`` from the genesis block.
-        """
-        chaindb = cls.get_chaindb_class()(db=base_db, genesis_config=genesis_config)
-        genesis_scoring = constantly(0)
+        genesis_scoring_class = sm_class.get_fork_choice_scoring_class()
+        genesis_score_class = genesis_scoring_class.get_score_class()
+        genesis_score = genesis_score_class.from_genesis(genesis_state, genesis_block)
```
Comment (ralexstokes) [inline eth2/beacon/chains/base.py]: the scoring used for the `genesis_block` requires the `genesis_state` in the score's `from_genesis` method so i merged the two functions (for genesis block and genesis state) into one function.

File: eth2/beacon/chains/base.py [hunk 2]
```diff
@@ -429,7 +426,9 @@ def import_block(
             raise ValidationError(
                 "Attempt to import block #{}.  Cannot import block {} before importing "
                 "its parent block at {}".format(
-                    block.slot, block.signing_root, block.parent_root
+                    block.slot,
+                    humanize_hash(block.signing_root),
+                    humanize_hash(block.parent_root),
```
Comment (ralexstokes) [inline eth2/beacon/chains/base.py]: just formatting on error message

File: eth2/beacon/db/chain.py [hunk 3]
```diff
@@ -412,18 +414,22 @@ def _get_block_signing_root_by_hash_tree_root(
...
+        db: DatabaseAPI, block_root: SigningRoot, score_class: Type[BaseScore]
+    ) -> BaseScore:
         try:
             encoded_score = db[SchemaV1.make_block_root_to_score_lookup_key(block_root)]
         except KeyError:
             raise BlockNotFound(
                 "No block with signing root {0} found".format(encode_hex(block_root))
             )
-        return ssz.decode(encoded_score, sedes=ssz.sedes.uint64)
+        return cast(BaseScore, score_class.deserialize(encoded_score))
```
Comment (ralexstokes) [inline eth2/beacon/db/chain.py]: `cast` required to satisfy `mypy`

File: eth2/beacon/db/chain.py [hunk 4]
```diff
@@ -437,31 +443,37 @@ def persist_block_chain(
...
         Return two iterable of blocks, the first containing the new canonical blocks,
         the second containing the old canonical headers
         """
         with self.db.atomic_batch() as db:
+            try:
+                first_block = next(iter(blocks))
+                if first_block.is_genesis:
+                    self._handle_exceptional_justification_and_finality(first_block)
+            except StopIteration:
+                pass
```
Comment (ralexstokes) [inline eth2/beacon/db/chain.py]: there was actually a bit of a bug here in that we were only checking for the genesis block in `persist_block` and *not* in `persist_block_chain`.

a future refactoring could unify this somewhat, but for now we just check in both places.

File: eth2/beacon/fork_choice/constant.py [hunk 5]
```diff
@@ -0,0 +1,29 @@
+from typing import Type
+
+from eth2.beacon.fork_choice.scoring import BaseForkChoiceScoring, BaseScore
+from eth2.beacon.types.blocks import BaseBeaconBlock
+from eth2.beacon.types.states import BeaconState
+
+
+class ConstantScoring(BaseForkChoiceScoring):
```
Comment (ralexstokes) [inline eth2/beacon/fork_choice/constant.py]: this class is a bit of plumbing/proxy to satisfy the scoring API

File: eth2/beacon/fork_choice/lmd_ghost.py [hunk 6]
```diff
@@ -21,6 +24,47 @@
...
+    def __eq__(self, other: Any) -> bool:
+        if not isinstance(other, LMDGHOSTScore):
+            return NotImplemented
+        return self._score == other._score
+
+    def serialize(self) -> bytes:
+        return ssz.encode(
+            self._score,
+            sedes=ssz.sedes.Vector(ssz.sedes.uint256, LMD_GHOST_SCORE_DATA_LENGTH),
+        )
```
Comment (ralexstokes) [inline eth2/beacon/fork_choice/lmd_ghost.py]: i like that we can now represent scores for LMD GHOST correctly, and not collapse to a single `int`.

File: eth2/beacon/chains/base.py [hunk 7]
```diff
@@ -236,21 +237,13 @@ def from_genesis(
...
-        genesis_block: BaseBeaconBlock,
-        genesis_config: Eth2GenesisConfig,
-    ) -> "BaseBeaconChain":
-        """
-        Initialize the ``BeaconChain`` from the genesis block.
-        """
-        chaindb = cls.get_chaindb_class()(db=base_db, genesis_config=genesis_config)
-        genesis_scoring = constantly(0)
+        genesis_scoring_class = sm_class.get_fork_choice_scoring_class()
+        genesis_score_class = genesis_scoring_class.get_score_class()
```
Comment (ralexstokes) [inline eth2/beacon/chains/base.py]: given that this is only place we are really using `get_fork_choice_scoring_class`, we could arguably collapse these two functions and only expose `get_score_class` on the `sm_class`.

File: eth2/beacon/tools/factories.py [hunk 8]
```diff
@@ -21,33 +25,68 @@ class BeaconChainFactory(factory.Factory):
...
+                if block.is_genesis:
+                    continue
+                # NOTE: ideally we use the ``import_block`` method
+                # on ``chain`` but for the time being we skip some
+                # validation corresponding to assumptions made in clients of
+                # this class. A future refactoring should use the external API.
+                chain.chaindb.persist_block(
+                    block, SerenityBeaconBlock, HigherSlotScoring()
+                )
+
```
Comment (ralexstokes) [inline eth2/beacon/tools/factories.py]: needed more control over genesis and how we bootstrap a `Chain` object in the factory during the syncing tests... so added some plumbing here to do this...

File: tests/libp2p/bcc/test_receive_server.py [hunk 9]
```diff
@@ -51,14 +51,16 @@ def import_block(
...
+            block, block.__class__, HigherSlotScoring()
         )
         return block, new_canonical_blocks, old_canonical_blocks
 
 
 async def get_fake_chain() -> FakeChain:
     genesis_config = Eth2GenesisConfig(MINIMAL_SERENITY_CONFIG)
     chain_db = AsyncBeaconChainDBFactory(genesis_config=genesis_config)
+    genesis_block = BeaconBlockFactory()
+    chain_db.persist_block(genesis_block, SerenityBeaconBlock, HigherSlotScoring())
```
Comment (ralexstokes) [inline tests/libp2p/bcc/test_receive_server.py]: took essentially this logic out of `AsyncBeaconChainDBFactory` so replicating the missing pieces here

File: tests/libp2p/bcc/test_syncing.py [hunk 10]
```diff
@@ -12,14 +14,20 @@
...
+    alice_chaindb = AsyncBeaconChainDBFactory()
+    bob_chaindb = AsyncBeaconChainDBFactory()
     peer_pair = ConnectionPairFactory(
-        alice_chaindb=alice_chaindb, bob_chaindb=bob_chaindb
+        alice_chaindb=alice_chaindb,
+        alice_branch=alice_branch,
+        bob_chaindb=bob_chaindb,
+        bob_branch=bob_branch,
+        genesis_state=genesis_state,
     )
```
Comment (ralexstokes) [inline tests/libp2p/bcc/test_syncing.py]: given the ordering of the objects here and the `asynccontextmanager` natuer of `ConnectionPairFactory`, had to do some modification to construct the genesis properly and in an order that set the rest of the assumptions used during syncing... like even having the genesis block (or our highest block) as the canonical head when resolving which blocks to sync

File: trinity/tools/bcc_factories.py [hunk 11]
```diff
@@ -108,19 +111,29 @@ class Meta:
...
+    genesis_state: BeaconState = None,
     cancel_token: CancelToken = None,
     handshake: bool = True,
 ) -> AsyncIterator[Tuple[Node, Node]]:
     if cancel_token is None:
         cancel_token = CancelTokenFactory()
-    alice_kwargs = {}
-    bob_kwargs = {}
+    alice_kwargs: Dict[str, Any] = {}
+    bob_kwargs: Dict[str, Any] = {}
```
Comment (ralexstokes) [inline trinity/tools/bcc_factories.py]: `mypy` needed explicit typing here

File: eth2/beacon/fork_choice/lmd_ghost.py [hunk 12]
```diff
@@ -21,6 +24,47 @@
...
 
+LMD_GHOST_SCORE_DATA_LENGTH = 2
+
+
+def score_block_by_root(root: SigningRoot) -> int:
+    return int.from_bytes(root, byteorder="big")
+
+
+class LMDGHOSTScore(BaseScore):
+    _score: Tuple[Gwei, int]
```
Comment (hwwhww) [inline eth2/beacon/fork_choice/lmd_ghost.py]: Perhaps add a comment like:
```suggestion
    # First score by the latest attesting balance, and then score by block root
    _score: Tuple[Gwei, int]  
```

File: eth2/beacon/fork_choice/lmd_ghost.py [hunk 13]
```diff
@@ -21,6 +24,47 @@
...
+            data, sedes=ssz.sedes.Vector(ssz.sedes.uint256, LMD_GHOST_SCORE_DATA_LENGTH)
+        )
+        return cls(score)
+
+    @classmethod
+    def from_genesis(
+        cls, genesis_state: BeaconState, genesis_block: BaseBeaconBlock
+    ) -> BaseScore:
+        score = (Gwei(0), score_block_by_root(SigningRoot(ZERO_HASH32)))
+        return LMDGHOSTScore(score)
```
Comment (hwwhww) [inline eth2/beacon/fork_choice/lmd_ghost.py]: nitpick, just notice that you use `return cls(score)` in `deserialize`.
```suggestion
        return cls(score)
```

File: eth2/beacon/fork_choice/lmd_ghost.py [hunk 14]
```diff
@@ -377,7 +417,7 @@ def on_attestation(
                     epoch=target.epoch, root=attestation.data.beacon_block_root
                 )
 
-    def scoring(self, block: BaseBeaconBlock) -> int:
+    def scoring(self, block: BaseBeaconBlock) -> LMDGHOSTScore:
         """
         Return the score of the target ``_block`` according to the LMD GHOST algorithm,
```
Comment (hwwhww) [inline eth2/beacon/fork_choice/lmd_ghost.py]:
```suggestion
        Return the score of the target ``block`` according to the LMD GHOST algorithm,
```

**Comment by ralexstokes** [comment]
yep, good point about comparing scores across forks -- this seems like a case where we can just extend the scoring metrics for each class to account for any special rules required at the fork!

thanks for the review!!

File: eth2/beacon/fork_choice/lmd_ghost.py (hunk 13 above)
Comment (ralexstokes) [reply eth2/beacon/fork_choice/lmd_ghost.py]: oh thank you! silly oversight on my part :)

File: eth2/beacon/fork_choice/lmd_ghost.py (hunk 14 above)
Comment (ralexstokes) [reply eth2/beacon/fork_choice/lmd_ghost.py]: thanks! this got lost in a prior edit

**Pull request merged**